### PR TITLE
Ensure floats are returned losslessly by the Rust ABI on 32-bit x86

### PR DESCRIPTION
**Pull request by beetrees**
Solves #115567 for the (default) `"Rust"` ABI. When compiling for 32-bit x86, this PR changes the `"Rust"` ABI to return floats indirectly instead of in x87 registers (with the exception of single `f32`s, which this PR returns in general purpose registers as they are small enough to fit in one). No change is made to the `"C"` ABI as that ABI requires x87 register usage and therefore will need a different solution.

### CODE REVIEW

File: compiler/rustc_ty_utils/src/abi.rs [hunk 1]
```diff
@@ -741,6 +741,38 @@ fn fn_abi_adjust_for_abi<'tcx>(
                 return;
             }
 
+            // Avoid returning floats in x87 registers on x86 as loading and storing from x87
+            // registers will quiet signalling NaNs.
+            if cx.tcx.sess.target.arch == "x86"
+                && arg_idx.is_none()
+                // Intrinsics themselves are not actual "real" functions, so theres no need to
+                // change their ABIs.
+                && abi != SpecAbi::RustIntrinsic
```
Comment (compiler-errors) [inline compiler/rustc_ty_utils/src/abi.rs:750]: This affects all ABIs, not just `Rust`?

Comment (compiler-errors) [reply compiler/rustc_ty_utils/src/abi.rs:750]: Ah, nvm, didn't see the outer nesting

**Review by compiler-errors — APPROVED** [review]
Changes lgtm, but I'm not an ABI expert

**Comment by compiler-errors** [comment]
so let's get another look at it 

r? compiler

**Comment by Nadrieril** [comment]
Not an ABI expert either

r? compiler

File: compiler/rustc_ty_utils/src/abi.rs [hunk 2]
```diff
@@ -741,6 +741,38 @@ fn fn_abi_adjust_for_abi<'tcx>(
...
+            {
+                match arg.layout.abi {
```
Comment (fmease) [inline compiler/rustc_ty_utils/src/abi.rs:752]: What about `F16`? Do we also need to account for it?

**Review by fmease — COMMENTED** [review]
Looks reasonable to me, too. I'm also not an expert on this matter ^^'.
@RalfJung, could you help us out?

File: compiler/rustc_ty_utils/src/abi.rs [hunk 3]
```diff
@@ -741,6 +741,38 @@ fn fn_abi_adjust_for_abi<'tcx>(
...
+                    // Handle similar to the way aggregates are handles below.
```
Comment (RalfJung) [inline compiler/rustc_ty_utils/src/abi.rs:753]: I don't understand this comment.

File: src/doc/rustc/src/platform-support.md [hunk 4]
```diff
@@ -204,6 +204,8 @@ target | std | notes
 
 [^x86_32-floats-x87]: Floating-point support on `i586` targets is non-compliant: the `x87` registers and instructions used for these targets do not provide IEEE-754-compliant behavior, in particular when it comes to rounding and NaN payload bits. See [issue #114479][x86-32-float-issue].
 
+[x86-32-float-issue]: https://github.com/rust-lang/rust/issues/114479]
```
Comment (RalfJung) [inline src/doc/rustc/src/platform-support.md:207]:
```suggestion
[x86-32-float-issue]: https://github.com/rust-lang/rust/issues/114479
```

**Comment by RalfJung** [comment]
> Looks reasonable to me, too. I'm also not an expert on this matter ^^'. @RalfJung, could you help us out?

My name appears in these issues as I gained just enough knowledge to identify that we have a problem, but I don't feel confident to approve any changes to our ABI, sorry.

With my limited knowledge however, this change does make a lot of sense. And it's a lot simpler than what I thought was necessary to do this. Very nice!

So... who *is* an ABI expert? @bjorn3, @nikic  and @eddyb are the first people coming to my mind.

(Maybe we should have some sort of dedicated ABI expert ping group. It would be really good to have some people in charge of ABI-related questions.)

**Comment by programmerjake** [comment]
I guess it can be left for a followup, but if SSE2 is available, I think `f32`/`f64` should be returned in `xmm0`, since that's where all the math is done anyway. This can be done by setting the `inreg` [LLVM parameter attribute](https://llvm.org/docs/LangRef.html#parameter-attributes) on the return type:

https://github.com/llvm/llvm-project/blob/575a6483062b8a77b35f48589b2acc1020195ac7/llvm/lib/Target/X86/X86CallingConv.td#L300-L304

**Comment by bjorn3** [comment]
I don't have time to review this for the next two weeks at least unfortunately.

**Review by workingjubilee — CHANGES_REQUESTED** [review]
Please add an assembly test that validates code with the intended calling convention is generated.

A behavioral test is not enough.

File: compiler/rustc_ty_utils/src/abi.rs (hunk 2 above)
Comment (workingjubilee) [reply compiler/rustc_ty_utils/src/abi.rs:752]: This should be an assembly test.

**Review by nikic — COMMENTED** [review]
I think a better approach to this would be to use the `fastcc` calling convention (note that this is different from x86_fastcallcc). One of the things it does is return floats in SSE registers, see https://llvm.godbolt.org/z/rjvhsGnqG. This will result in better codegen, and I think is generally in the spirit of the Rust calling convention (that is, do whatever is fast for the target, without any ABI guarantees).

**Comment by RalfJung** [comment]
Cc @tgross35 as you mentioned non-SSE-ABI issues elsewhere

**Comment by RalfJung** [comment]
@nikic there has been an answer to your question, do you have other concerns related to this PR?

File: tests/ui/abi/numbers-arithmetic/return-float.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
+//@ run-pass
+//@ compile-flags: -Copt-level=0
+// FIXME(#114479): LLVM miscompiles loading and storing `f32` and `f64` when SSE is disabled.
+// There's no compiletest directive to ignore a test on i586 only, so just always explicitly enable
+// SSE2.
+//@ compile-flags: -Ctarget-feature=+sse2
```
Comment (nikic) [inline tests/ui/abi/numbers-arithmetic/return-float.rs:6]: Isn't this going to fail on non-x86 hosts?

File: tests/assembly/x86-return-float.rs [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+// There's no compiletest directive to ignore a test on i586 only, so just always explicitly enable
+// SSE2.
+// Use the same target CPU as `i686` so that LLVM orders the instructions in the same order.
+//@ compile-flags: -Ctarget-feature=+sse2 -Ctarget-cpu=pentium4
+// Force frame pointers to make ASM more consistent between targets
+//@ compile-flags: -O -C force-frame-pointers
+//@ filecheck-flags: --implicit-check-not fld --implicit-check-not fst
+//@ revisions: unix windows
+//@[unix] ignore-windows
+//@[windows] only-windows
```
Comment (nikic) [inline tests/assembly/x86-return-float.rs:13]: Just to check, is the relevant difference here windows or msvc? If the latter, then compiletest sets up MSVC and NONMSVC prefixes for you.

**Review by workingjubilee — APPROVED** [review]
Thanks!

**Pull request merged**